runtime/runtime/src/congestion_control.rs

```diff
@@ -335,9 +408,97 @@ impl ReceiptSinkV2<'_> {
+        else {
+            return Ok(None);
+        };
+
+        let mut requests = Vec::new();
+        for shard_id in self.outgoing_buffers.shards() {
+            if let Some(request) =
+                self.generate_bandwidth_request(shard_id, trie, side_effects, &params)?
+            {
+                requests.push(request);
+            }
+        }
+
+        Ok(Some(BandwidthRequests::V1(BandwidthRequestsV1 { requests })))
+    }
+
+    fn generate_bandwidth_request(
+        &self,
+        to_shard: ShardId,
+        trie: &dyn TrieAccess,
+        side_effects: bool,
+        params: &BandwidthSchedulerParams,
+    ) -> Result<Option<BandwidthRequest>, StorageError> {
+        let Some(first_receipt) =
+            self.outgoing_buffers.get_first_receipt(to_shard, trie, side_effects)?
+        else {
+            // No receipts in the outgoing buffer
+            return Ok(None);
+        };
+        let first_receipt_size: u64 =
+            receipt_size(&first_receipt).expect("First receipt size doesn't fit in u64");
+
+        let shard_u8: u8 = to_shard.try_into().expect("ShardId doesn't fit into u8");
+
+        // We start collecting metadata information after the protocol upgrade,
+        // but it can take a moment for the older receipts to be removed from
+        // the buffer and the metadata to become fully initialized.
+        // Metadata contains only information about the receipts added after the
+        // protocol upgrade. We can't generate a request based on the metadata
+        // if doesn't cover all of the receipts stored in the buffer.
+        let is_metadata_fully_initialized = first_receipt.queue_metadata_version()
+            == Some(QueueMetadataVersion::OutgoingBufferMetadataV0);
+
+        let request = if is_metadata_fully_initialized {
+            let metadata = self.outgoing_buffer_metadatas.metadatas.get(&to_shard).expect(
+                "Outgoing buffer metadata should exist for shard with fully initialized metadata",
+            );
+            // Collect sizes of receipts stored in this outgoing buffer.
+            // We need to start with the size of the first receipt. It could happen that
+            // the size of the first group is larger than max_receipt_size, and it could be that
+            // num_shards * base_bandwidth + first_group_size > max_shard_bandwidth, in which case
+            // we would never be able to grant the bandwidth required to send the first group.
+            // It's an unlikely scenario, but to protect against it we always request bandwidth for the first receipt,
+            // it's always possible to grant enough bandwidth for one receipt.
+            let mut receipt_group_sizes = metadata.iter_receipt_group_sizes();
+            let first_group_size = receipt_group_sizes.next().unwrap();
+            let receipt_sizes = [first_receipt_size, first_group_size - first_receipt_size]
```

Comment:
nit: It's a bit strange that the bandwidth request is made based on receipt sizes but you pass in the group sizes. Maybe make it be based on groups and just pass the first receipt size as an additional argument to the ctor? I'm not sure if this suggestion is any good, feel free to ignore.

Reply:
BandwidthRequest would ideally operate on receipts, but groups are an optimization that has to be done. But it's a bit unexact, I'll add a TODO and revisit it later, there's enough going on in this PR.

Reply:
I tried to find a better name, but it's pretty hard. Bandwidth request can be created from receipts or receipt groups, or prefix sums of receipt sizes. `receipt_sizes` is generic enough to cover all these cases.
If you have an idea for a better name, I can change it.